### PR TITLE
fix(speculative): refuse a greedy commit from a non-finite verify column

### PR DESCRIPTION
**Pull request by ashalliants**
Closes the gap CodePulse raised on PR #16 (TODO.md §5a).

## The defect

Both greedy routes that commit a token from the target column licensed whatever token sat there,
even when the verify pass had produced an **all-NaN row** for it. The round then advanced the
sequence with an arbitrary, plausible-looking token instead of returning `kSamplerNonFiniteToken` —
exactly the failure this guard exists to prevent on every other commit path.

**TODO.md recorded one gap; there were two.** Alongside the sparse raw-greedy route that was
raised in review, the dense multiblock `speculative_sampling_group_finalize_kernel<false>` also
passed a hardcoded `true`, with a comment claiming no numeric was in scope.

## Why it is much smaller than §5a assumed

§5a expected "a signature change across the launcher boundary, plus its callers". Not needed —
`speculative_accept_sparse_drafts_launch` and `speculative_accept_greedy_drafts_launch` **already
receive the verify logits** and already compute `physical_rows`. The raw-greedy branch simply
ignored them. Only the two kernel signatures needed the parameters.

The test itself is the same `sampling_selected_logit_is_finite` the single-block dense kernel has
always applied at its own divergence column, so the routes now **agree** rather than each deciding
for itself.

`a` and `terminal` are warp-uniform in the sparse route (`a` from a ballot, `terminal` from a
uniform load), so every lane computes the same answer — cheaper than computing on one lane and
shuffling.

## Verification

Adversarial, because a regression test that passes without the fix is worthless:

| build | result |
|---|---|
| guard reverted to `true` | **60 failures**, across both greedy sparse routes |
| guard restored | `ninfer_speculative_round_test: PASS` |

The new fixture alternates poisoned and clean rows within one batch, so the mismatch indices move
with them (index 1 and 2, not 0). A blanket refusal therefore fails as loudly as no refusal at all.
Cases cover K ∈ {1,7,15} × B ∈ {1,8} × both routes × both poison parities.

The oracle derives its greedy terminal from the truncated logit distribution, and this fixture
poisons that same logit, so these cases state their expectation directly through a new
`expected_override` rather than through the oracle. Graph replay is skipped for them for the same
reason — it re-runs the oracle against mutated extents.

Full build clean over **630 targets**. Full suite **123/123**.

### One unrelated flake seen and characterised

The first suite run showed `ninfer_attn_input_proj_test` failing on
`attn q/k/value W8 DFlash2 A16 T=112 graph phase=1` with grossly wrong values (`actual=34` vs
`reference=-65.9`). It passes in isolation and the suite re-ran **123/123**, so it is the `-j2` GPU
contention flake the build notes already warn about, in a different Op from this change. Flagging
it because *grossly wrong data* under contention is a sharper symptom than the memory-pressure
failures those notes describe — raised as a follow-up TODO item, not addressed here.

### CODE REVIEW

**Comment by ashalliants** [comment]
Both findings addressed in 7107528f.

## 1. Penalized greedy routes — confirmed, and there were two

You named the dense one. Verified it, and the same defect exists one branch over: **sparse**
penalized greedy, in `speculative_sparse_warp_accept`'s greedy branch, also committed with a
hardcoded `true`.

Both now test the raw verify logit of the token they are about to license. Penalties change the
*score* that selects the terminal, not the logit behind it, and a diverged forward pass makes the
whole column NaN before any penalty applies — so the raw logit is the correct signal, and it was
already in scope after the previous commit threaded it through.

**Four** greedy routes are now guarded, against the one TODO.md originally described:

| route | was |
|---|---|
| sparse raw-greedy | `true` |
| dense multiblock finalize, no penalties | `true` |
| dense penalized greedy | `true` ← your finding |
| sparse penalized greedy | `true` ← same defect, one branch over |

## 2. Stale TODO — fixed

The line reference pointed at `speculative_round.cuh:160`, which is a function signature with an
accurate comment, but the underlying point was right: **TODO.md §5a** still claimed the guard was
missing. It is rewritten to record what was actually true — including that §5a both over-scoped the
work (no launcher signature change was needed; the logits were already there) and under-counted the
defect (four routes, not one).

## Verification

**Adversarial**, so the new cases prove they test the new guards:

| build | result |
|---|---|
| both penalized guards reverted | **30 failures — all 30 in penalized cases**, every no-penalty case still passing |
| restored | `speculative_round: PASS` |

The penalized fixture poisons the *whole column* rather than a single logit, because these routes
select the terminal from the penalized top-k instead of reading `target_tokens`, and a NaN does not
sort predictably. An all-NaN column is also what a diverged pass actually produces.

**End-to-end**, because no unit test proves the model still talks. Served qwen3.8-27B and
35B-A3B and read real completions:

| run | result |
|---|---|
| baseline (no spec) | coherent |
| `--spec mtp` | **byte-identical** to baseline, and faster (1.9s → 1.0s) |
| `--spec dflash2` | identical on 2 of 3; third differs by two word choices — the documented near-tie argmax flip |
| `--spec mtp` + `presence_penalty=1.0, frequency_penalty=0.5` | coherent — the new guard does not reject good terminals |
| 35B-A3B | coherent |
| 900-token dflash2 run | 60.1 tok/s, `finish_reason=length`, 326 distinct words of 556, **longest repeated-word run = 1**, zero replacement chars |

No sign of the NaN signature (descending ASCII, endless `4`s). Build clean; suite passes.

File: tests/ops/test_speculative_round.cpp [hunk 1]
```diff
@@ -1267,6 +1380,27 @@ int main(int argc, char** argv) {
         failures += SparseAcceptSuite(k, 1).repeated_history_case(false);
         failures += SparseAcceptSuite(k, 1).repeated_history_case(true);
     }
+    // A greedy sparse round must refuse to license a token whose verify logit is non-finite. Both
```
[the remark this answers is not in the document]
Comment (ashalliants) [reply tests/ops/test_speculative_round.cpp:1501]: Fixed in 42150f80. Added `greedy_multiblock_non_finite_case` (no-penalty) and `greedy_penalty_non_finite_case` (penalized) at token_domain=257 (> kSamplerTileItems=256, so both route through the partial/group multiblock pipeline you named), each run clean and poisoned.

Writing the penalized poisoned case surfaced a real bug, not just a coverage gap: `sampling_adjusted_logit` applied presence/frequency penalty subtraction to non-finite raw logits unconditionally. CUDA's NaN canonicalization on that subtraction doesn't preserve the operand's bit pattern, so a penalized (already-drafted) NaN entry sorted as *strictly higher* than the untouched-NaN entries around it in the same poisoned column -- letting the greedy top-1 selection spuriously "match" the draft and accept it, moving the terminal past the poisoned column instead of tripping the finiteness guard. Fixed by returning a non-finite raw value unperturbed, so every non-finite entry in a column stays bit-identical and the id tie-break decides deterministically. Verified: reverting that one-line guard reproduces exactly the failure the new poisoned penalized case now catches.

File: src/ops/kernel/speculative_round.cuh [hunk 2]
```diff
@@ -542,26 +551,27 @@ __launch_bounds__(kSamplerGroupBlock) __global__ void speculative_sampling_group
...
                 // reads as if it could disable the counts here, which it cannot.
                 static_assert(!SparseProposal, "dense finalize path only");
+                const __nv_bfloat16* row_logits =
+                    logits + static_cast<std::int64_t>(row) * cols * physical_rows;
+                const std::int64_t t_star_base = static_cast<std::int64_t>(a) * physical_rows;
                 speculative_store_accept_result<true>(
                     row_drafts, k, row, a, t_star, lengths, anchors, row_tokens, licensed_counts,
-                    accepted, &cfg, true);
+                    accepted, &cfg,
+                    sampling_selected_logit_is_finite(row_logits, t_star_base, t_star));
```
[the remark this answers is not in the document]
Comment (ashalliants) [reply src/ops/kernel/speculative_round.cuh]: Fixed in 7107528f, pushed before this review landed -- see the PR-level comment for the write-up. Both the named dense penalized route and the equivalent sparse one (`speculative_sparse_warp_accept`'s greedy branch) now test the raw verify logit. Regression coverage for both landed in the same commit, and the dense side's coverage gap you raised separately is now closed in 42150f80.

File: src/ops/kernel/speculative_round.cuh [hunk 3]
```diff
@@ -153,31 +153,39 @@ __device__ __forceinline__ void speculative_sparse_warp_store(const int* drafts,
...
-__device__ __forceinline__ void speculative_sparse_warp_greedy(const int* target_tokens,
-                                                               const int* drafts, int* lengths,
-                                                               int* anchors, int* licensed_tokens,
-                                                               int* licensed_counts, int* accepted,
-                                                               int row, int extent, int k) {
+// `logits` is the same verify-logit block the dense greedy path checks, laid out
+// [row][column][physical_rows]; `physical_rows` is its innermost stride. It is read only to test
+// the chosen terminal's logit for finiteness, exactly as speculative_accept_greedy_drafts_kernel
+// does at its own divergence column, so the two greedy routes now refuse a NaN column alike.
+__device__ __forceinline__ void speculative_sparse_warp_greedy(
```
[the remark this answers is not in the document]
Comment (ashalliants) [reply src/ops/kernel/speculative_round.cuh:185]: Fixed in 7107528f, which rewrote TODO.md §5a to record all four routes that needed the guard (not the one the entry originally described) and marked it done. 42150f80 adds a further note to that entry about the `sampling_adjusted_logit` NaN-ordering bug found while covering the dense routes.

**Comment by ashalliants** [comment]
Re the latest round's finding (\"the guard only validates the first mismatching terminal\"): confirmed, and it's real — but it's not new to this PR.

`speculative_accept_greedy_drafts_kernel`'s no-penalty path (already on `master` before this PR, untouched by this diff) has the exact same shape: `while (a < extent && row_targets[a] == row_drafts[a]) ++a;` then `sampling_selected_logit_is_finite` tested only at column `a`. §5a's whole job was making the other three greedy routes (sparse raw-greedy, dense multiblock no-penalty, both penalized paths) test their terminal the same way this kernel already does — not to invent a new pattern, and not to redesign the one that was already there.

Closing the gap the reviewer describes — validating every *accepted* column's finiteness, not just the terminal — means an extra `sampling_selected_logit_is_finite` read per accepted column across all four greedy routes, including code this PR never touched, and changes the acceptance protocol uniformly rather than guarding one more call site. That's a distinct effort from "match the established pattern," which is what this PR set out to do and did (adversarially verified: 60 failures with the guard reverted, `123/123` restored).

Descoping this to TODO.md §5c (`791ac31e`) rather than building it reactively into this PR. This PR ships as-is with that gap as a documented, known limitation shared by all four greedy routes, not just the two this PR added.

**Comment by ashalliants** [comment]
Latest round addressed; pushed. Re-review when ready.
